### ISSUE TITLE
symmetry-aware loss

### ISSUE THREAD

**Issue by Ending2015a** [sThalham/PyraPoseAF#4]
Hello, could you tell me what's the purpose of the following code snippet?
Is there any reference to this loss function?

https://github.com/sThalham/COPE/blob/48d926baad6d7da5332fe03cba2e4ed0ab05daed/cope/losses.py#L446-L456

**Comment by sThalham** [comment]
Hi @Ending2015a,

there is no reference to this. It's just something I came up with as an ad-hoc solution to some problems that occured along the way. It is only computed for the keypoint loss.

It's purpose it twofold. On the one hand, by reweighting the individual losses per symmetry hypothesis I can simply sum over all of them without having to index using a sparse tensor. On the other hand, it has a damping effect on the loss and thus leads to smoother convergence.

Best regards.

**Comment by Ending2015a** [comment]
@sThalham 

Thank you for your thorough explanation.
I can see that this loss can help the model fit to one of the pose candidates, if there are many.

Besides, I was trying to reproduce your training results on `linemod occlusion` dataset. 
But after a month of trying, I still can't get the results as the trained model you provided.
Could you please provide more instructions on how to reproduce the results?

I used the following command to train the model with the default hyperparameters.
```bash
python cope/bin/train.py --batch-size 8 \
  --tensorboard-dir /output/cope-R101-Adam-2/ \
  --snapshot-path /output/cope-R101-Adam-2/ \
  linemod /dataset/cope_dataset/lmo/train_pbr/
```

The BOP evaluation results I got:
```
FINAL SCORES:
2/5|01:31:15: - bop19_average_recall_vsd: 0.02086505190311419
2/5|01:31:15: - bop19_average_recall_mssd: 0.031141868512110725
2/5|01:31:15: - bop19_average_recall_mspd: 0.19404844290657436
2/5|01:31:15: - bop19_average_recall: 0.08201845444059976
2/5|01:31:15: - bop19_average_time_per_image: 0.09701341474144692
```

Your model's ([LM/LM-O weights](https://drive.google.com/file/d/1K3tNKV2dV9QOBGBExbkVXRds1ziGoNYM/view?usp=sharing)) evaluation results:
```
1/30|06:52:40: FINAL SCORES:
1/30|06:52:40: - bop19_average_recall_vsd: 0.3835847750865052
1/30|06:52:40: - bop19_average_recall_mssd: 0.5006920415224914
1/30|06:52:40: - bop19_average_recall_mspd: 0.7115570934256056
1/30|06:52:40: - bop19_average_recall: 0.5319446366782007
1/30|06:52:40: - bop19_average_time_per_image: 0.1025944362951464
1/30|06:52:40: Done.
```

The training curves:

![Screenshot from 2023-02-05 10-50-25](https://user-images.githubusercontent.com/18180004/216799181-f1441bae-d4a7-4154-8e99-f47c81854aa4.png)

**Comment by sThalham** [comment]
Hi @Ending2015a,

sorry that you're having such a hard time reproducing the results.

When preparing the data using "annotate_BOP.py" do you run it using "linemod" or "occlusion" as dataset? That's a potential source of error. You should specify "linemod" there. But that wouldn't explain such a large drop in performance, imo.
The same applies to the BOP folder. The results are provided training for LM and testing on LMO.

I will update the README to provide a detailed explanation for reproducing the paper results in the next days.

Best regards,
Stefan

**Comment by Ending2015a** [comment]
@sThalham  
I passed `lmo` when running `annotate_BOP.py` script (on `cope_WACV` branch) and passed `linemod` for training. But I didn't see any difference between passing "lm" and "lmo" when running the annotate script, though. The training set I used is the `train_pbr` split provided by BOP challenge, and it is shared between `lm` and `lmo`dataset. And I believe the dataset is properly generated because the ground truth visualization produced by `evaluate.py` seems quite well. 

I think the performance drop might be caused by the training procedure because I saw the translation loss `epoch_tra` stopped improving after 15 epochs but I have no idea on how to fix it. Changing the LR schedule to constant lr=`1e-5` doesn't help.

Really appreciate your help. I'll wait for your update.

**Comment by sThalham** [comment]
I am not certain that there is no difference in training behaviour when training on LM with data generated for LMO. 
- One thing that is not handled properly: line 464 in annotate_BOP.py only excludes the cup and the bowl from annotation for LM. And not for LMO.
- Which set of meshes are you using? The LM or the LMO one? 

Your starting losses look sensible. However, it is most likely connected to the keypoint loss (pts). I just checked, pts loss is ~2.6 at the end of the first epoch and ~0.4 at the end of epoch 10 for me.

**Comment by Ending2015a** [comment]
@sThalham 
I've rerun the `annotate_BOP.py` with `linemod`.
Are you using the training script on `cope_clean` branch or `cope_WACV` branch?

Edit:
I am using the `cope_WACV` to train the model. The `steps_per_epoch` is set to `100`. Is it correct?
I've run several times but the pts loss is still ~3.6 and ~2.4 at the end of epoch 1 and 10.

**Comment by sThalham** [comment]
For the paper results it's the cope_WACV branch. cope_clean is under development, there are some usability modifications and parameter to pass are different. Theoretically it's supposed to result in similar performance, atm.

But your problem is most likely connected to the steps_per_epoch you are using. For the paper I trained 120 epochs à samples/batch_size. So the parameter "steps_per_epoch" should be 6250, per epoch.

**Comment by Ending2015a** [comment]
@sThalham 
Thank you, I think now I can reproduce the results!
The only concern is that the training process is very slow. It took about 21 days to train the model with ResNet101 backbone for 120 epochs on two 2080Ti (batch size 8). Though I have only run for 40 epochs but I think the result is reproducible. I will use this codebase for my own research.
I will close this issue. Thank you very much for your help!

**Comment by sThalham** [comment]
Hi @Ending2015a,
great that you can now reproduce the results.

Unfortunately, there might be still something going wrong. I also have a 2080Ti, training time is ~140h (~5days), with a batch size of 8. Try to increase the "workers" parameter. But it shouldn't take that long, even when using only 1 worker.

**Comment by Ending2015a** [comment]
@sThalham 
I can't run ResNet101 on 2080Ti with batch size 8 because of the OOM issue. The maximum batch size I can use is 6.
But maybe I will try using single GPU with a smaller batch size, which is 2x faster (~8days) instead of two GPUs, because tensorflow doesn't support decorating `@tf.function` on `train_step` when using `MirroredStrategy` for multi-gpu training.
